include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +389,531 @@ class BidirectionalNode
+    auto Comparator = [&P](auto const &Edge) { return Edge.Neighbor == &P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
+  }
+
+public:
+  SuccessorEdgeIterator findSuccessorEdge(DerivedType const &S) {
+    return SuccessorEdgeIterator(findSuccessorImpl(S),
+                                 SuccessorFilters::ToView);
+  }
+  ConstSuccessorEdgeIterator findSuccessorEdge(DerivedType const &S) const {
+    return ConstSuccessorEdgeIterator(findSuccessorImpl(S),
+                                      SuccessorFilters::ToConstView);
+  }
+  PredecessorEdgeIterator findPredecessorEdge(DerivedType const &S) {
+    return PredecessorEdgeIterator(findPredecessorImpl(S),
+                                   PredecessorFilters::ToView);
+  }
+  ConstPredecessorEdgeIterator findPredecessorEdge(DerivedType const &S) const {
+    return ConstPredecessorEdgeIterator(findPredecessorImpl(S),
+                                        PredecessorFilters::ToConstView);
+  }
+
+  SuccessorIterator findSuccessor(DerivedType const &S) {
+    return SuccessorIterator(findSuccessorImpl(S),
+                             SuccessorFilters::ToNeighbor);
+  }
+  ConstSuccessorIterator findSuccessor(DerivedType const &S) const {
+    return ConstSuccessorIterator(findSuccessorImpl(S),
+                                  SuccessorFilters::ToConstNeighbor);
+  }
+  PredecessorIterator findPredecessor(DerivedType const &S) {
+    return PredecessorIterator(findPredecessorImpl(S),
+                               PredecessorFilters::ToNeighbor);
+  }
+  ConstPredecessorIterator findPredecessor(DerivedType const &S) const {
+    return ConstPredecessorIterator(findPredecessorImpl(S),
+                                    PredecessorFilters::ToConstNeighbor);
+  }
+
+public:
+  bool hasSuccessor(DerivedType const &S) const {
+    return findSuccessorImpl(S) != Successors.end();
+  }
+  bool hasPredecessor(DerivedType const &S) const {
+    return findPredecessorImpl(S) != Predecessors.end();
+  }
+
+public:
+  size_t successorCount() const { return Successors.size(); }
+  size_t predecessorCount() const { return Predecessors.size(); }
+  size_t neighborCount() const {
+    return Successors.size() + Predecessors.size();
+  }
+
+  bool hasSuccessors() const { return Successors.size() != 0; }
+  bool hasPredecessors() const { return Predecessors.size() != 0; }
+  bool hasNeighbors() const { return hasSuccessors() || hasPredecessors(); }
+
+protected:
+  using OwnerIteratorImpl = typename EdgeOwnerContainer::const_iterator;
+  using ViewIteratorImpl = typename EdgeViewContainer::const_iterator;
+
+  auto removeSuccessorImpl(OwnerIteratorImpl InputIterator) {
+    auto Iterator = Successors.begin();
+    std::advance(Iterator,
+                 std::distance<OwnerIteratorImpl>(Iterator, InputIterator));
+
+    if (Successors.empty())
+      return Iterator;
+
+    // Maybe we should do some extra checks as to whether `Iterator` is valid.
```

Comment:
@aleclearmind what do you think? What should we check? Something like `Iterator != Container.end()` doesn't sound like a bad idea, anything else?

Reply:
Works.
If possible, we could have `Container.begin() <= Iterator < Container.end()`?
But usually I don't try to be too much smarter than the typical guarantees you get from STL.

Reply:
I'm kind of reluctant of using `operator<` with iterators. This could cause weird problems if we ever decide to switch the underlying container away from a vector-like. But yeah, I don't think there's anything better, let's go with this one.

Reply:
Again, sounds like overzealous checks. Up to you.